### PR TITLE
Error report

### PR DESCRIPTION
**Pull request by ahuoguo**
This PR implements an error-reporting tool into Verus to simplify the process of reporting unexpected Verus errors/bugs. It essentially takes a “snapshot” of a given file – including the file itself, all of the file’s dependencies, the user’s Z3 version and Verus version, the command line arguments, and stdout/stderr information – and compresses it into a zip file that preserves the original file hierarchy. All relevant information is recorded into a .toml file that is zipped in with the dependencies. The created zip file (named YYYY-MM-DD-hh-mm-ss.zip) will be saved to the current working directory.

This PR puts the flag `--record` into Verus itself. With this PR, if I am in directory `bar` and want to report a Verus bug that I am experiencing in `foo.rs`, I could call in the command line `verus --record foo.rs`. This would produce a zip file in the directory `bar` that includes `foo.rs`, any dependency of `foo.rs`, and a file called `error_report.toml` that would look something like the following:

```
report-schema-version = "1"
title = "Error report file - details and dependencies"

[command-line-arguments]
args = "verus rust_verify/example/playground.rs"

[verification-results]
encountered-vir-error = false
errors = 0
success = true
verified = 3

[versions]
z3-version = """
Z3 version 4.12.2 - 64 bit
"""

[versions.verus-version]
commit = "07f6e9959e4398c67a2307e4bc8123662580960a"
profile = "debug"
toolchain = "1.68.0-aarch64-apple-darwin"
version = "0.2023.07.14.07f6e99"

[versions.verus-version.platform]
arch = "aarch64"
os = "macos"

[verus-output]
stderr = """
...
"""

[verus-output.stdout.times-ms]
...

[verus-time]
verus-time = 0.58403575
```

Usage: `verus --record INPUT [options]`

### CODE REVIEW

File: source/error_report/src/main.rs
Comment (jaybosamiya) [inline source/error_report/src/main.rs:188]: Potential old comment left behind?

File: source/error_report/src/main.rs
Comment (jaybosamiya) [inline source/error_report/src/main.rs:191]: Actually allow the zip file to compress things, rather than just store things
```suggestion
        .compression_method(zip::CompressionMethod::Bzip2)
```

File: source/error_report/src/main.rs
Comment (jaybosamiya) [inline source/error_report/src/main.rs:209]: Isn't this exactly equivalent to `fs::read_to_string(filepath)`? Not sure why this is a separate function

File: source/rust_verify/src/main.rs
Comment (jaybosamiya) [inline source/rust_verify/src/main.rs:121]: This expects `error_report` to have been built with debug-only; might want to consider checking whether the debug or release builds are around, and pick the right one.

File: source/rust_verify/src/main.rs
Comment (jaybosamiya) [inline source/rust_verify/src/main.rs:124]: Out of curiosity, why is this comment here? I'd assume `current_dir` to be set to the current dir anyways (which is what this comment implies)

File: source/rust_verify/src/util.rs
Comment (jaybosamiya) [inline source/rust_verify/src/util.rs:156]: Not sure why this diff showed up, should be cleaned up

**Review by jaybosamiya — COMMENTED** [review]
Thanks @elaustell and @ahuoguo; this is a good first draft for the error reporting functionality. I've added a bunch of comments inline that should be addressed, before we consider merging this (also please make sure to see the "hidden comments" that GitHub collapses down too)

Also, the main PR message says "Note: this error-report relies on an old version of --version flag. This issue will be addressed after --version flag is incorporated."; is this still the case, or has the `--version` version (heh) been fixed up.

Thanks again! And looking forward to have `--error-report` merged in soon!

File: source/error_report/src/main.rs
Comment (elaustell) [reply source/error_report/src/main.rs:48]: Maybe but the output itself doesn't have to do with the dep info, this output relates specifically to what verus outputs to stdout and stderr when run on the given file. So I'm not sure the name verus_dep_info_output is accurate.

Comment (jaybosamiya) [reply source/error_report/src/main.rs:56]: Oh indeed; ignore this comment.

File: source/error_report/Cargo.toml
Comment (jaybosamiya) [reply source/error_report/Cargo.toml]: Ah, I see! Well this is now replaced out with the `vargo build` changes, so I guess it is resolved either way.

File: source/error_report/src/main.rs
Comment (jaybosamiya) [inline source/error_report/src/main.rs:29]: Now that the expected call is `verus --error-report <args..>`, it should probably be reflected here:
```suggestion
        println!("Usage: verus --error-report <args..>");
```

File: source/error_report/src/main.rs
Comment (jaybosamiya) [inline source/error_report/src/main.rs:33]: I assume this is guaranteed to be passed in via the implementation of the `--error-report` flag, but that is still an assumption being made, so prob should be documented as a comment

File: source/error_report/src/main.rs
Comment (jaybosamiya) [inline source/error_report/src/main.rs:38]: `args[2..]` is placed into `our_args`; maybe reuse that via a `clone` or such?

File: source/error_report/src/main.rs
Comment (jaybosamiya) [reply source/error_report/src/main.rs:33]: On further thought, maybe this should be passed in as an environment variable? For example `VERUS_HOME` or such?

File: source/rust_verify/src/config.rs
Comment (jaybosamiya) [inline source/rust_verify/src/config.rs:269]:
```suggestion
        "Zip relevant files triggering a Verus error to aid in reporting it",
```

File: source/rust_verify/src/main.rs
Comment (jaybosamiya) [inline source/rust_verify/src/main.rs:138]: This probably should be handled by the `error_report` binary?

**Review by jaybosamiya — COMMENTED** [review]
Some more comments added inline

File: source/error_report/src/main.rs
Comment (jaybosamiya) [reply source/error_report/src/main.rs:34]: Fair, I think it is fine to leave it as an argument here

File: source/error_report/src/main.rs
Comment (jaybosamiya) [reply source/error_report/src/main.rs]: that's somewhat surprising, but I think a `String` is fine for now

File: source/rust_verify/src/main.rs
Comment (jaybosamiya) [reply source/rust_verify/src/main.rs]: The intention with placing the `--time` removal in the `error_report` binary is that we want to report (to the `error_report` binary) something that fully captures the command the user used for Verus, as closely as possible; the fact that an `--error-report` flag was passed is only a way to invoke the `error_report` binary; but whether the user _explicitly_ passed in `--time` or not _may_ potentially be something that we might want/need to actually track (for `--time` I guess it is not _as_ crucial, but hopefully you see my reasoning on why the two flags are somewhat different in their removal)

File: source/rust_verify/src/main.rs
Comment (jaybosamiya) [inline source/rust_verify/src/main.rs:149]:
```suggestion
                "error_report executable not found; try running `vargo build --release`?"
```

File: source/rust_verify/src/main.rs
Comment (jaybosamiya) [inline source/rust_verify/src/main.rs:137]: Copying over comments from https://github.com/verus-lang/verus/pull/612#discussion_r1254769157 which is now marked as outdated:

@jaybosamiya:
> This probably should be handled by the `error_report` binary?

@ahuoguo:
> I removing `--error-report` flag at the same time, so these two might want to stay in the same level?

@jaybosamiya:
> The intention with placing the `--time` removal in the `error_report` binary is that we want to report (to the `error_report` binary) something that fully captures the command the user used for Verus, as closely as possible; the fact that an `--error-report` flag was passed is only a way to invoke the `error_report` binary; but whether the user _explicitly_ passed in `--time` or not _may_ potentially be something that we might want/need to actually track (for `--time` I guess it is not _as_ crucial, but hopefully you see my reasoning on why the two flags are somewhat different in their removal)

**Review by jaybosamiya — COMMENTED** [review]
Thanks for making the updates.

Modulo above minor comments above (for the message in a panic, and the place for `--time`), I think the PR is in reasonable-enough shape for folks to start trying out on `main`. I still think some more cleanup can be done, but imho that shouldn't block this PR from being merged and being used by folks, so will do that at a separate point in time.  I think I can approve the PR once those changes are made.

Wrt merging itself, I think we should get at least an approval from @utaal since this touches command line arguments and such.  `source/error_report` doesn't actually impact regular Verus usage, so the total scale of changes to review should be fairly minimal.

Sidenote: when squash-merging this PR (after approvals), feel free to drop me off the co-author list; my changes in the changeset were fairly minor fixes here.

**Comment by utaal** [comment]
Thank you for working on this! I have a few high-level comments and some specifics.

1. I wasn't completely comfortable with having to emit a `.d` file and having to find it by name, but I've had a look at what rustc libraries offer and it seems there isn't a straightforward way to obtain this information otherwise (I believe they're computed here: https://github.com/verus-lang/verus/blob/main/dependencies/rustc_interface/src/passes.rs#L549-L662) We might want to try and ask rustc for the `.d` output path (via `OutputFilenames::path`), instead of guessing it but this isn't critical.

2. Generally speaking, we should only use `.expect` (on `Option` and `Result`) for cases we never expect to happen (and `.unwrap` only for trivial cases, e.g. when you've just otherwise checked that an `Option` is `Some`). For example, invoking `verus` with `--error-report` and an invalid command line argument results in a panic (because `error_report` cannot find the `.d` file.)

3. We cannot expect that `error_report` is at `../../target/(debug|release)/error_report` relative to the `verus` path, the regular `target` can get clobbered by successive builds (in fact, this fails if you run `vargo run -p rust_verify -- --error-report rust_verify/example/playground.rs`). The only stable directory is `verus-target/(debug|release)` (which gets wiped by vargo if it becomes stale or incomplete). `vargo` should copy `error_report` into `verus-target/(debug|release)/.`, like it does for, e.g. `rust_verify` (https://github.com/verus-lang/verus/blob/main/tools/vargo/src/main.rs#L674-L675); then `verus` can find it via the `verus_root`, computed here https://github.com/verus-lang/verus/blob/main/source/rust_verify/src/main.rs#L91

**Comment by utaal** [comment]
4. The way rustc collects dependencies may be surprising (i.e. files mentioned by a macro); shall we (a.) gracefully handle a case in which we cannot find one of the files, and (b.) print a list of files included in the archive? Also, I think it would be good to mention which information is included in the archive (in brief) in the output, so that the user knows what they're about to submit (they can obviously go ahead and check the zip file themselves).

5. The outputs for `time` and `version` are currently just text, which would make it harder to parse later. If we use `--output-json` (which I've recently updated for this purpose) we can include its output directly in the toml, ideally by parsing the json and re-emitting it as toml. This way the toml file can be parsed directly by a tool. (This may sound tricky but should be relatively straightforward; let me know if I can help).

File: source/error_report/src/main.rs
Comment (utaal) [inline source/error_report/src/main.rs:36]: This may not be the version of `z3` used by `verus`, as it depends on whether `VERUS_Z3_PATH` is set. We can have `verus` pass this information along to `error-report`.

File: source/error_report/src/main.rs
Comment (utaal) [inline source/error_report/src/main.rs:200]: Can we construct the format we want, instead of just truncating the output of `to_string` and replacing characters?

File: source/error_report/src/main.rs
Comment (utaal) [inline source/error_report/src/main.rs:210]: Why does this need to be set explicitly? What happens on Windows?

**Comment by utaal** [comment]
Also, thoughts on naming this `--package-report`, or similar? (It isn't necessarily only useful for errors, and it'd be good to communicate that it "packages" the report, that the user is in charge of sending/uploading.)

File: source/error_report/src/main.rs
Comment (utaal) [inline source/error_report/src/main.rs:79]: If we panic between the time these files are created and when this line is reached, they won't be cleaned up (this has happened to me while trying this out). Error handling with `Result` should help with this.

**Comment by parno** [comment]
> > Also, thoughts on naming this `--package-report`, or similar? (It isn't necessarily only useful for errors, and it'd be good to communicate that it "packages" the report, that the user is in charge of sending/uploading.)
> 
> Shed by the lights of Jon, I suggest `--report`

`--report` could still reasonably be interpreted as doing the reporting, i.e., sending information off to a Verus server somewhere.  `--package-report` makes it clearer that the option is just assembling the report, not doing anything with it.  If we want something more concise, perhaps `--collate` or `--collect`?

**Comment by utaal** [comment]
Recording offline conversation: @ahuoguo, @jaybosamiya, and @parno suggested `--record`. That sounds good to me.

File: source/error_report/README.md
Comment (utaal) [inline source/error_report/README.md:1]: I'd suggest replacing this with a section in the guide. It's unlikely that users will find it here.

File: source/error_report/src/main.rs
Comment (utaal) [inline source/error_report/src/main.rs:26]: Not a big deal, but the idiomatic rust to do is probably for this to be `Option<String>`. Then your check later can be:

```rust
let Some(file_path) = file_path else { return Err(...); };
```

File: source/error_report/src/main.rs
Comment (utaal) [inline source/error_report/src/main.rs:84]:
```suggestion
    // mandating --time and --output-json flag, we remove both flags here to prevent passing duplicates to verus
```

File: source/error_report/src/main.rs
Comment (utaal) [inline source/error_report/src/main.rs:177]:
```suggestion
                "failed to parse verification results with error message {}",
```

File: source/error_report/src/main.rs
Comment (utaal) [inline source/error_report/src/main.rs:184]: Like above, the failure here is parsing the json, not pretty-printing.

File: source/rust_verify/src/config.rs
Comment (utaal) [inline source/rust_verify/src/config.rs:273]:
```suggestion
        "Rerun verus and package source files of the current crate to the current directory, alongside with output and version information. The file will be named YYYY-MM-DD-HH-MM-SS.zip. If you are reporting an error, please keep the original arguments in addition to this flag",
```

File: source/rust_verify/src/config.rs
Comment (utaal) [inline source/rust_verify/src/config.rs:167]:
```suggestion
    const OPT_RECORD: &str = "record";
```

File: source/rust_verify/src/config.rs
Comment (utaal) [inline source/rust_verify/src/config.rs:412]:
```suggestion
        error_report: matches.opt_present(OPT_RECORD),
```

File: source/rust_verify/src/config.rs
Comment (utaal) [inline source/rust_verify/src/config.rs:272]:
```suggestion
        OPT_RECORD,
```

File: source/rust_verify/src/config.rs
Comment (utaal) [inline source/rust_verify/src/config.rs:74]:
```suggestion
    pub record: bool,
```

File: source/rust_verify/src/main.rs
Comment (utaal) [inline source/rust_verify/src/main.rs:130]:
```suggestion
    if our_args.record {
```

File: source/rust_verify/src/main.rs
Comment (utaal) [inline source/rust_verify/src/main.rs:153]: Don't fail silently, either panic (which I think is okay here, as you can expect a non-broken verus build) or print a message that says what went wrong.

**Review by utaal — APPROVED** [review]
Thank you for addressing the comments.
I have added a couple minor comments/suggestions. Please address those, then feel free to squash and merge. (If GitHub suggests adding me as a co-author because of the suggestions, please remove me.)

File: source/error_report/src/main.rs
Comment (jaybosamiya) [inline source/error_report/src/main.rs:113]:
```suggestion
                format!("failed to delete `.d` file with this error message: {}", x)
```

File: source/error_report/src/main.rs
Comment (jaybosamiya) [inline source/error_report/src/main.rs:129]:
```suggestion
                format!("failed to delete `.d` file with this error message: {}", x)
```

**Review by jaybosamiya — APPROVED** [review]
Thanks for all the updates; some minor remaining suggestions (all of which should just be easily accepted suggestions). Feel free to squash-and-merge once you get a CI green after making the updates based on the suggestions.

**Pull request merged**
